File: Antlr/src/main/scala/io/github/chaosunity/casc/parsing/class/Function.scala
package io.github.chaosunity.casc.parsing.`class`

import io.github.chaosunity.casc.parsing.`type`.Type
import io.github.chaosunity.casc.parsing.expression.FunctionParameter
import io.github.chaosunity.casc.parsing.statement.Statement

import java.util

class Function(val name: String,
               val returnType: Type,
               val parameters: util.List[FunctionParameter],
               val rootStatement: Statement) {
}

File: src/main/kotlin/io/github/chaosunity/casc/parsing/Function.kt
package io.github.chaosunity.casc.parsing

import io.github.chaosunity.casc.bytecode.MethodFactory
import io.github.chaosunity.casc.parsing.node.statement.Statement
import io.github.chaosunity.casc.parsing.scope.AccessModifier
import io.github.chaosunity.casc.parsing.scope.FunctionSignature
import io.github.chaosunity.casc.parsing.type.Type

open class Function<T>(
    val signature: FunctionSignature,
    val rootStatement: Statement<*>,
    val accessModifier: AccessModifier,
    val static: Boolean = false
) where T : Function<T> {
    val name: String = signature.name
    open val returnType: Type = signature.returnType

    fun generate(factory: MethodFactory) =
        factory.generate(this)
}